lib/core/build/classes/datawrapper/Chart.php

```diff
@@ -144,70 +146,46 @@ protected function getDataFilename() {
     }
 
     /**
-     * writes any asset to the file system store
+     * writes any asset to the asset store using v3 api
      */
     public function writeAsset($filename, $data) {
-        $cfg = $GLOBALS['dw_config'];
+        if (!$this->assets) $this->assets = [];
 
-        if (isset($cfg['charts-s3'])
-          && isset($cfg['charts-s3']['write'])
-          && $cfg['charts-s3']['write'] == true) {
+        [$status, $body] = call_v3_api('PUT',
+            '/charts/' . $this->getId() . '/assets/' . $filename, $data, "text/csv");
 
-            $config = $cfg['charts-s3'];
+        $this->assets[$filename] = $data;
 
-            $filenamePath = 's3://' . $cfg['charts-s3']['bucket'] . '/' .
-                $this->getRelativeDataPath() . '/' . $filename;
-
-            file_put_contents($filenamePath, $data);
-        } else {
-            $path = $this->getDataPath();
-
-            if (!file_exists($path)) {
-                mkdir($path, 0775);
-            }
-
-            $filenamePath = $path . '/' . $filename;
-
-            file_put_contents($filenamePath, $data);
-        }
         $this->setLastModifiedAt(time());
-        return $filenamePath;
     }
 
     /**
-     * writes raw csv data to the file system store
+     * writes raw csv data to the asset store using v3 api
      *
      * @param csvdata  raw csv data string
      */
     public function writeData($csvdata) {
-        return $this->writeAsset($this->getDataFilename(), $csvdata);
+        $this->writeAsset($this->getDataFilename(), $csvdata);
     }
 
     /**
      * load any asset from file system
      */
     public function loadAsset($filename) {
-        $config = $GLOBALS['dw_config'];
-
-        if (isset($config['charts-s3']) &&
-            $config['charts-s3']['read']) {
+        if (!$this->assets) $this->assets = [];
 
-            $s3url = 's3://' . $config['charts-s3']['bucket'] . '/' .
-              $this->getRelativeDataPath() . '/' . $filename;
+        if (empty($this->assets[$filename])) {
+            [$status, $body] = call_v3_api('GET',
+            '/charts/' . $this->getId() . '/assets/' . $filename);
 
-            try {
-                return file_get_contents($s3url);
-            } catch (Exception $ex) {
-                return '';
-            }
-        } else {
-            $filenamePath = $this->getDataPath() . '/' . $filename;
-            if (!file_exists($filenamePath)) {
-                return '';
+            if ($status == "404") {
+                $this->assets[$filename] = "";
             } else {
-                return file_get_contents($filenamePath);
+                $this->assets[$filename] = gettype($body) == "string" ? $body : json_encode($body, 1);
```

Comment:
same here, I think we should check for a `200` success code to catch other server errors. otherwise this code could try to json encode error messages etc.

Reply:
also, why aren't we using `json_encode_safe` here?

Reply:
didn't know / forgot about `json_encode_safe`, makes sense though